src/rpc/rpc.cc

```diff
@@ -3,323 +3,294 @@
  * \file rpc/rpc.cc
  * \brief Implementation of RPC utilities used by both server and client sides.
  */
-#include "./rpc.h"
-
-#include <csignal>
 #if defined(__linux__)
-#include <unistd.h>
-#endif
+#include "./rpc.h"
 
-#include <dgl/runtime/container.h>
-#include <dgl/runtime/parallel_for.h>
-#include <dgl/packed_func_ext.h>
 #include <dgl/array.h>
+#include <dgl/packed_func_ext.h>
 #include <dgl/random.h>
+#include <dgl/runtime/container.h>
+#include <dgl/runtime/parallel_for.h>
 #include <dgl/zerocopy_serializer.h>
-#include "../runtime/resource_manager.h"
+#include <tensorpipe/tensorpipe.h>
+#include <unistd.h>
+
+#include <csignal>
+#include <future>
+
 #include "../c_api_common.h"
+#include "../runtime/resource_manager.h"
 
 using dgl::network::StringPrintf;
 using namespace dgl::runtime;
 
 namespace dgl {
 namespace rpc {
 
+using namespace tensorpipe;
+
 RPCStatus SendRPCMessage(const RPCMessage& msg, const int32_t target_id) {
-  std::shared_ptr<std::string> zerocopy_blob(new std::string());
-  StreamWithBuffer zc_write_strm(zerocopy_blob.get(), true);
-  zc_write_strm.Write(msg);
-  int32_t nonempty_ndarray_count = zc_write_strm.buffer_list().size();
-  zerocopy_blob->append(reinterpret_cast<char*>(&nonempty_ndarray_count),
-                        sizeof(int32_t));
-  network::Message rpc_meta_msg;
-  rpc_meta_msg.data = const_cast<char*>(zerocopy_blob->data());
-  rpc_meta_msg.size = zerocopy_blob->size();
-  rpc_meta_msg.deallocator = [zerocopy_blob](network::Message*) {};
-  CHECK_EQ(RPCContext::ThreadLocal()->sender->Send(
-    rpc_meta_msg, target_id), ADD_SUCCESS);
-  // send real ndarray data
-  for (auto ptr : zc_write_strm.buffer_list()) {
-    network::Message ndarray_data_msg;
-    ndarray_data_msg.data = reinterpret_cast<char*>(ptr.data);
-    if (ptr.size == 0) {
-      LOG(FATAL) << "Cannot send a empty NDArray.";
-    }
-    ndarray_data_msg.size = ptr.size;
-    NDArray tensor = ptr.tensor;
-    ndarray_data_msg.deallocator = [tensor](network::Message*) {};
-    CHECK_EQ(RPCContext::ThreadLocal()->sender->Send(
-      ndarray_data_msg, target_id), ADD_SUCCESS);
-  }
+  RPCContext::ThreadLocal()->sender->Send(msg, target_id);
```

Comment:
with this, can we send messages in multiple threads?

Reply:
You're right. Cannot do this now since the context is `ThreadLocal`.

Reply:
Updated. Now it supports multithread with unit test